### PR TITLE
feat(store): `GetNoteAuthenticationInfo` endpoint

### PR DESCRIPTION
**Pull request by polydez**
Resolves: https://github.com/0xPolygonMiden/miden-node/issues/395

This PR adds `GetNoteAuthenticationInfo` endpoint. Prior this we had to use `GetNotesById` endpoint for getting inclusion proofs, but it returns a lot of unnecessary information for such task.

Also `GetBlockInputs` endpoint now returns not only found requested note IDs, but also their inclusion proofs.

### CODE REVIEW

File: proto/note.proto [hunk 1]
```diff
@@ -23,6 +23,11 @@ message Note {
     optional bytes details = 6;
 }
 
+message NoteInclusionProof {
+    digest.Digest note_id = 1;
+    merkle.MerklePath merkle_path = 2;
+}
```
Comment (bobbinth) [inline proto/note.proto:29]: This needs to include a bit more data. Specifically, note inclusion proof should be sufficient to prove that a note is present in the chan. This requires:

1. A Merkle path from the leaf to the root of note Merkle tree of a block.
2. Index of the note in this block.
3. Block's note root + sub hash (or alternatively, the entire block header).
4. MMR path to the block in the current chain MMR.

Since several notes could be from the same block, it may make sense to de-duplicate the data. Though, maybe it doesn't matter too much and at least for now we can send all required data on per note basis.

File: proto/rpc.proto [hunk 2]
```diff
@@ -10,6 +10,7 @@ service Api {
     rpc GetAccountDetails(requests.GetAccountDetailsRequest) returns (responses.GetAccountDetailsResponse) {}
     rpc GetBlockByNumber(requests.GetBlockByNumberRequest) returns (responses.GetBlockByNumberResponse) {}
     rpc GetBlockHeaderByNumber(requests.GetBlockHeaderByNumberRequest) returns (responses.GetBlockHeaderByNumberResponse) {}
+    rpc GetNoteInclusionProofs(requests.GetNotesByIdRequest) returns (responses.GetNoteInclusionProofsResponse) {}
```
Comment (bobbinth) [inline proto/rpc.proto:13]: Even if the requests are the same - we should probably create a different request message (i.e., we probably shouldn't reuse `GetNotesByIdRequest` request here).

File: proto/store.proto [hunk 3]
```diff
@@ -14,6 +14,7 @@ service Api {
     rpc GetBlockByNumber(requests.GetBlockByNumberRequest) returns (responses.GetBlockByNumberResponse) {}
     rpc GetBlockHeaderByNumber(requests.GetBlockHeaderByNumberRequest) returns (responses.GetBlockHeaderByNumberResponse) {}
     rpc GetBlockInputs(requests.GetBlockInputsRequest) returns (responses.GetBlockInputsResponse) {}
+    rpc GetNoteInclusionProofs(requests.GetNotesByIdRequest) returns (responses.GetNoteInclusionProofsResponse) {}
```
Comment (bobbinth) [inline proto/store.proto:17]: Same comment as above.

**Review by bobbinth — COMMENTED** [review]
Thank you! Looks good. This is not a full review yet - but I left a couple of comments inline. The main one is that note inclusion proof is actually a bit more complex than just a single Merkle path.

File: proto/note.proto (hunk 1 above)
Comment (polydez) [reply proto/note.proto:29]: @bobbinth thank you for clarification!
> 4. MMR path to the block in the current chain MMR.
Do we need this field?
I found implementation of `NoteInclusionProof` in `miden-base` and there is no such field:
https://github.com/0xPolygonMiden/miden-base/blob/f80276e93521f896b2ff5544f42ddaba6e50038e/objects/src/notes/origin.rs#L23-L35

File: crates/store/src/db/sql.rs [hunk 4]
```diff
@@ -665,6 +665,58 @@ pub fn select_notes_by_id(conn: &mut Connection, note_ids: &[NoteId]) -> Result<
...
+
+        let note_id_data = row.get_ref(1)?.as_blob()?;
+        let note_id = NoteId::read_from_bytes(note_id_data)?;
+
+        let batch_index = row.get(2)?;
+        let note_index = row.get(3)?;
+        let node_index_in_block = BlockNoteIndex::new(batch_index, note_index).to_absolute_index();
+
+        let merkle_path_data = row.get_ref(4)?.as_blob()?;
+        let merkle_path = MerklePath::read_from_bytes(merkle_path_data)?;
```
Comment (Mirko-von-Leipzig) [inline crates/store/src/db/sql.rs:710]: Meta: something to consider is adding typed getters using an extension trait for the sqlite row. We did something similar in pathfinder [here](https://github.com/eqlabs/pathfinder/blob/7ec71dcbc35372a4424f4772b3a02ecfe7e19c5c/crates/storage/src/params.rs#L151) which I think added quite a lot in terms of consistency and readability.

**Review by Mirko-von-Leipzig — APPROVED** [review]
LGTM

File: crates/block-producer/src/batch_builder/batch.rs [hunk 5]
```diff
@@ -88,7 +85,7 @@ impl TransactionBatch {
     #[instrument(target = "miden-block-producer", name = "new_batch", skip_all, err)]
     pub fn new(
         txs: Vec<ProvenTransaction>,
-        found_unauthenticated_notes: BTreeMap<NoteId, MerklePath>,
+        found_unauthenticated_notes: BTreeMap<NoteId, NoteInclusionProof>,
```
Comment (bobbinth) [inline crates/block-producer/src/batch_builder/batch.rs:88]: Related to the previous comments: I don't think `BTreeMap<NoteId, NoteInclusionProof>` would be sufficient to capture all required information. We'll probably need to create a separate struct which contains all additional info (i.e., block headers and MMR proofs).

The questions is what to call this struct. One option is `UnauthenticatedNoteWitness` - but I don't love this name.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good. Not a full review but I did leave some comments inline. The main one is I think just `NoteInclusionProof` objects are not sufficient we are missing the MMR part of the proofs.

File: proto/rpc.proto [hunk 6]
```diff
@@ -12,6 +12,7 @@ service Api {
     rpc GetAccountStateDelta(requests.GetAccountStateDeltaRequest) returns (responses.GetAccountStateDeltaResponse) {}
     rpc GetBlockByNumber(requests.GetBlockByNumberRequest) returns (responses.GetBlockByNumberResponse) {}
     rpc GetBlockHeaderByNumber(requests.GetBlockHeaderByNumberRequest) returns (responses.GetBlockHeaderByNumberResponse) {}
+    rpc GetNoteInclusionProofs(requests.GetNoteInclusionProofsRequest) returns (responses.GetNoteInclusionProofsResponse) {}
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply proto/rpc.proto]: Removed.

File: proto/responses.proto [hunk 7]
```diff
@@ -119,6 +119,11 @@ message GetNotesByIdResponse {
     repeated note.Note notes = 1;
 }
 
+message GetNoteInclusionProofsResponse {
+    // Lists note inclusion proofs returned by the database
+    repeated note.NoteInclusionProof proofs = 1;
+}
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply proto/responses.proto]: I've updated this to include block MMR proofs.

File: proto/responses.proto [hunk 8]
```diff
@@ -88,8 +88,8 @@ message GetBlockInputsResponse {
     // The requested nullifiers and their authentication paths
     repeated NullifierBlockInputRecord nullifiers = 4;
 
-    // The list of requested notes which were found in the database
-    repeated digest.Digest found_unauthenticated_notes = 5;
+    // The list of requested unauthenticated notes with their inclusion proofs which were found in the database
+    repeated note.NoteInclusionProof found_unauthenticated_notes = 5;
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply proto/responses.proto:92]: I've ignored this for now, I'll update this once the other endpoint's api is set.

File: crates/store/src/db/sql.rs (hunk 4 above)
Comment (Mirko-von-Leipzig) [reply crates/store/src/db/sql.rs:710]: Created #442

File: crates/block-producer/src/batch_builder/batch.rs (hunk 5 above)
Comment (Mirko-von-Leipzig) [reply crates/block-producer/src/batch_builder/batch.rs:88]: Ignored for now

File: crates/store/src/db/mod.rs [hunk 9]
```diff
@@ -224,6 +224,24 @@ impl Db {
             })?
     }
 
+    /// Loads multiple block headers from the DB.
+    #[instrument(target = "miden-store", skip_all, ret(level = "debug"), err)]
+    pub async fn select_many_block_headers(
```
Comment (Mirko-von-Leipzig) [inline crates/store/src/db/mod.rs:229]: Or maybe `multiple_` instead? I also considered overloading the `select_block_headers` to take an optional vec instead.

File: crates/store/src/state.rs [hunk 10]
```diff
@@ -397,6 +403,51 @@ impl State {
...
+        // We also avoid accessing the db in the block as this would delay
+        // dropping the guard.
+        let merkle_paths = {
+            let state = self.inner.read().await;
+            let forest = state.chain_mmr.forest();
+
+            blocks
+                .iter()
+                .map(|block_num| {
+                    let proof = state.chain_mmr.open(todo!(), todo!())?.merkle_path;
```
Comment (Mirko-von-Leipzig) [inline crates/store/src/state.rs:431]: How do I convert an absolute index into a tree + offset index as required by `open`?

My brain tells me this is simple; but for some reason I'm glitching here. Probably just some modulo variation? Or does one have to go bit searching?

File: crates/store/src/types.rs [hunk 11]
```diff
@@ -1,2 +1,12 @@
...
 pub type BlockNumber = u32;
 pub type AccountId = u64;
+
+/// Data required to verify a block's MMR proof.
+pub struct BlockMmrProof {
+    pub block_number: BlockNumber,
+    pub mmr_path: MerklePath,
+    pub sub_hash: Digest,
+    pub note_root: Digest,
+}
```
Comment (Mirko-von-Leipzig) [inline crates/store/src/types.rs:12]: Is this living here okay?

File: crates/store/src/state.rs [hunk 12]
```diff
@@ -397,6 +403,54 @@ impl State {
...
 
+    /// Queries all the note inclusion proofs matching a certain Note IDs from the database.
+    pub async fn get_note_inclusion_proofs(
+        &self,
+        note_ids: Vec<NoteId>,
+    ) -> Result<BTreeMap<NoteId, NoteInclusionProof>, DatabaseError> {
+        self.db.select_block_note_inclusion_proofs(note_ids).await
+    }
+
+    pub async fn get_block_mmr_proofs(
```
Comment (bobbinth) [inline crates/store/src/state.rs:414]: I would rename this into `get_block_inclusion_proofs()`.

File: crates/proto/src/domain/mod.rs [hunk 13]
```diff
@@ -9,18 +9,20 @@ pub mod transactions;
...
 
-pub fn convert<T, From, To>(from: T) -> Vec<To>
+pub fn convert<T, From, To, R>(from: T) -> R
 where
     T: IntoIterator<Item = From>,
     From: Into<To>,
+    R: FromIterator<To>,
 {
     from.into_iter().map(|e| e.into()).collect()
 }
```
Comment (bobbinth) [inline crates/proto/src/domain/mod.rs:19]: Similar to one of the previous comments, maybe this could be something like:
```rust
pub fn convert<From, To>(from: impl IntoIterator<Item = From>) -> impl Iterator<Item = To>
where
    From: Into<To>
```
This way, the caller would be responsible for calling `.collect()` or could use the returned iterator in some other ways.

Though, I'm wondering if we are getting much convenience from this function at all (and maybe it make sense to remove it). Though, this would probably be for another PR.

File: crates/proto/src/domain/notes.rs [hunk 14]
```diff
@@ -1,17 +1,21 @@
...
+    crypto::hash::rpo::RpoDigest,
+    notes::{NoteId, NoteInclusionProof, NoteMetadata, NoteTag, NoteType},
     Felt,
 };
 
-use crate::errors::{ConversionError, MissingFieldHelper};
+use crate::{
+    errors::{ConversionError, MissingFieldHelper},
+    generated::note::{NoteInclusionProof as NoteInclusionProofPb, NoteMetadata as NoteMetadataPb},
+};
```
Comment (bobbinth) [inline crates/proto/src/domain/notes.rs:10]: nit (and not for this PR): I think it would be cleaner to add `Proto` suffix instead of `Pb` to the aliases. For example, `NoteMetadata as NoteMetadataProto`.

File: crates/block-producer/src/store/mod.rs [hunk 15]
```diff
@@ -56,15 +55,15 @@ pub trait Store: ApplyBlock {
...
     ///
-    /// TODO: right now this return only Merkle paths per note, but this will need to be updated to
-    /// return full authentication info.
+    /// TODO: right now this returns only Merkle paths per note, but this will need to be updated to
+    ///       return full authentication info.
     async fn get_note_authentication_info(
         &self,
         notes: impl Iterator<Item = &NoteId> + Send,
-    ) -> Result<BTreeMap<NoteId, MerklePath>, NotePathsError>;
+    ) -> Result<BTreeMap<NoteId, NoteInclusionProof>, NotePathsError>;
```
Comment (bobbinth) [inline crates/block-producer/src/store/mod.rs:66]: Maybe we can use a new struct for the return type here - something like:
```rust
pub struct NoteAuthenticationInfo {
    block_proofs: Vec<BlockInclusionProof>,
    note_proofs: BTreeMap<NoteId, NoteInclusionProof>,
}
```
This can the also be used in the `BlockInputs` struct.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left some comments inline (and I think answered all outstanding questions - but let me know if I missed something).

File: crates/store/src/state.rs (hunk 10 above)
Comment (bobbinth) [reply crates/store/src/state.rs:431]: Opened https://github.com/0xPolygonMiden/crypto/pull/324 to improve this in the future.

File: crates/proto/src/domain/mod.rs (hunk 13 above)
Comment (Mirko-von-Leipzig) [reply crates/proto/src/domain/mod.rs:19]: I suspect removal and handing control back to the caller ito allocations and `iter().map(..)` chaining makes the most sense. Though I don't have the full sense of the ergonomics here yet.

File: crates/proto/src/domain/notes.rs (hunk 14 above)
Comment (Mirko-von-Leipzig) [reply crates/proto/src/domain/notes.rs:10]: I'll see if it doesn't make more sense to alias the module instead

```rust
use crate::generated::note as proto;

impl From<T> for proto::T {...}
```

File: crates/store/src/server/api.rs [hunk 16]
```diff
@@ -255,6 +256,54 @@ impl api_server::Api for StoreApi {
...
+        request: tonic::Request<GetNoteInclusionProofsRequest>,
+    ) -> Result<Response<GetNoteInclusionProofsResponse>, Status> {
+        info!(target: COMPONENT, ?request);
+
+        let note_ids = request.into_inner().note_ids;
+
+        let note_ids: Vec<RpoDigest> = try_convert(note_ids)
+            .map_err(|err| Status::invalid_argument(format!("Invalid NoteId: {}", err)))?;
+
+        let note_ids: Vec<NoteId> = note_ids.into_iter().map(From::from).collect();
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply crates/store/src/server/api.rs:278]: #449

File: crates/store/src/db/sql.rs [hunk 17]
```diff
@@ -718,6 +770,31 @@ pub fn select_block_header_by_block_num(
...
+    let mut result = Vec::with_capacity(blocks.len());
+
+    for block in blocks {
+        let header: Vec<u8> = stmt
+            .query_row([block], |row| row.get(0))
+            .optional()?
+            .ok_or(DatabaseError::BlockNotFoundInDb(block))?;
+        let header = BlockHeader::read_from_bytes(&header)?;
+        result.push(header);
+    }
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply crates/store/src/db/sql.rs:793]: You're right, but I had concerns with the output not aligning with the input since the input is a slice and not a set.

And selecting using `IN` would return a set.

I'll see if I can't improve this and the function parameters.

File: crates/store/src/server/api.rs [hunk 18]
```diff
@@ -255,6 +256,54 @@ impl api_server::Api for StoreApi {
...
+        // Massage into shape required by protobuf
+        let note_proofs = note_proofs.iter().map(Into::into).collect();
+        let block_proofs = block_proofs
+            .into_iter()
+            .map(|proof| BlockMmrProof {
+                block_num: proof.block_number,
+                sub_hash: Some(proof.sub_hash.into()),
+                note_root: Some(proof.note_root.into()),
+                mmr_path: Some((&proof.mmr_path).into()),
+            })
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply crates/store/src/server/api.rs:301]: Do you mean the `proto/domain` module, or the store's `state` module? If the former then we also need to move this type into `proto` or `miden-objects`.f

I'm going to assume the latter for now.

File: crates/proto/src/domain/notes.rs (hunk 14 above)
Comment (Mirko-von-Leipzig) [reply crates/proto/src/domain/notes.rs:10]: Ended up using `generated` for the most part.

File: crates/store/src/db/sql.rs [hunk 19]
```diff
@@ -665,6 +665,58 @@ pub fn select_notes_by_id(conn: &mut Connection, note_ids: &[NoteId]) -> Result<
...
+/// Select note inclusion proofs matching the NoteId, using the given [Connection].
+///
+/// # Returns
+///
+/// - Empty map if no matching `note`.
+/// - Otherwise, note inclusion proofs, which `note_id` matches the `NoteId` as bytes.
+pub fn select_block_note_inclusion_proofs(
+    conn: &mut Connection,
+    note_ids: &[NoteId],
+) -> Result<BTreeMap<NoteId, NoteInclusionProof>> {
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply crates/store/src/db/sql.rs:677]: I've changed the input to `BTreeSet` instead, due to concerns given [here](https://github.com/0xPolygonMiden/miden-node/pull/421#discussion_r1723223858).

File: crates/store/src/db/mod.rs [hunk 20]
```diff
@@ -307,6 +325,25 @@ impl Db {
             })?
     }
 
+    /// Loads all the note inclusion proofs matching a certain Note IDs from the database,
+    /// grouped by block.
+    #[instrument(target = "miden-store", skip_all, ret(level = "debug"), err)]
+    pub async fn select_block_note_inclusion_proofs(
+        &self,
+        note_ids: Vec<NoteId>,
+    ) -> Result<BTreeMap<NoteId, NoteInclusionProof>> {
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply crates/store/src/db/mod.rs:334]: https://github.com/0xPolygonMiden/miden-node/pull/421#discussion_r1723223858

File: crates/block-producer/src/store/mod.rs (hunk 15 above)
Comment (Mirko-von-Leipzig) [reply crates/block-producer/src/store/mod.rs:66]: I'm busy implementing this; but I'm starting to feel like we should have split this PR up much more.

e.g. add this functionality to the store api, and only handle consuming it in the block producer in a separate PR.

File: crates/proto/src/domain/notes.rs (hunk 14 above)
Comment (bobbinth) [reply crates/proto/src/domain/notes.rs]: I actually prefer adding type aliases as I think it makes the code a bit easier to read - e.g., instead of `generated::merkle::MerklePath` we get `MerklePathProto`. But let's keep things as they currently are in this PR.

File: crates/store/src/db/sql.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply crates/store/src/db/sql.rs:797]: You mean that `blocks` parameter could contain duplicate block numbers? I think we should think of this as an error - though not sure we should check for it (at least at this level).

I'm also not sure we should enforce that for every passed-in block number there must be a record in the DB. So, I think this is OK that this function returns a vector of different size as compared to the passed-in vector?

Basically, the way I think of this function is that given a list of block numbers, it returns a set of matching block headers from the database. But it can also return no results if none block headers match.

[the remark this answers is not in the document]
Comment (bobbinth) [reply crates/store/src/db/sql.rs:797]: And to add: it would then be a responsibility of the caller to make sense of these results.

File: crates/store/src/db/sql.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply crates/store/src/db/sql.rs]: One potential concern I have with using `BTreeMap` and `BTreeSet` where we can avoid them is performance (inserting into these should be a bit more expensive than inserting into a vector). But maybe that's a case of premature optimization.

But in general, I'm not sure we need to align the input/output types here - there is no guarantee here that for every passed-in note ID we'll get an inclusion proof back. So, I think it is fine to pass in a slice of notes and get back a vector of results (possibility of different length and in different order). But maybe I'm missing something?

File: crates/store/src/server/api.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply crates/store/src/server/api.rs]: I meant moving `BlockMmrProof` (or what is now `BlockInclusionProof`) into `proto/domain` as I assumed we'll need to use it in the `block-producer` crate as well. But if it is only used here, we can keep it in this module.

File: crates/store/src/state.rs [hunk 21]
```diff
@@ -87,6 +94,21 @@ struct InnerState {
     account_tree: SimpleSmt<ACCOUNT_TREE_DEPTH>,
 }
 
+/// Data required to verify a block's inclusion proof.
+pub struct BlockInclusionProof {
+    pub block_header: BlockHeader,
+    pub mmr_path: MerklePath,
+}
+
+impl From<BlockInclusionProof> for miden_node_proto::generated::block::BlockInclusionProof {
```
Comment (bobbinth) [inline crates/store/src/state.rs:103]: Following one of the above comments: I think it would be a bit easier to read this line if it was:
```rust
impl From<BlockInclusionProof> for BlockInclusionProofProto
```
or
```rust
impl From<BlockInclusionProof> for proto::BlockInclusionProof
```

File: crates/block-producer/src/store/mod.rs (hunk 15 above)
Comment (bobbinth) [reply crates/block-producer/src/store/mod.rs:66]: I'm fine with splitting this into multiple PRs (I too find that it has become a bit difficult to follow) and doing `store` part in one PR and `block-producer` part in another makes sense to me.

File: crates/store/src/state.rs (hunk 21 above)
Comment (Mirko-von-Leipzig) [reply crates/store/src/state.rs:103]: Ah yeah my `grep` missed this one. Thanks.

File: crates/store/src/db/sql.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply crates/store/src/db/sql.rs]: I think I just got lost along the way and forgot that the input and output set are already misaligned ie not all inputs have an output. I'll change it to vectors when I get a chance again.

File: crates/block-producer/src/batch_builder/batch.rs [hunk 22]
```diff
@@ -141,10 +139,11 @@ impl TransactionBatch {
...
                     // except the nullifier)
-                    found_unauthenticated_notes
-                        .get(&input_note_header.id())
-                        .map(|_path| InputNoteCommitment::from(input_note.nullifier()))
-                        .unwrap_or_else(|| input_note.clone())
+                    if found_unauthenticated_notes.contains_note(&input_note_header.id()) {
+                        InputNoteCommitment::from(input_note.nullifier())
+                    } else {
+                        input_note.clone()
+                    }
```
Comment (Mirko-von-Leipzig) [inline crates/block-producer/src/batch_builder/batch.rs:146]: What isn't clear to me is where/what the note inclusion proof should be used for?

File: crates/block-producer/src/test_utils/block.rs [hunk 23]
```diff
@@ -22,7 +22,14 @@ pub async fn build_expected_block_header(
...
 ) -> BlockHeader {
-    let last_block_header = *store.last_block_header.read().await;
+    let last_block_header = *store
+        .block_headers
+        .read()
+        .await
+        .iter()
+        .max_by_key(|(block_num, _)| *block_num)
+        .unwrap()
+        .1;
```
Comment (Mirko-von-Leipzig) [inline crates/block-producer/src/test_utils/block.rs:32]: This is really awkward to do everywhere, but I didn't think its safe to extract this as a method since it holds a lock. So calling it elsewhere could easily result in an accidental dead-lock.

**Review by Mirko-von-Leipzig — COMMENTED** [review]
I've rebased and I *think* did the block producer portion as well.

File: proto/store.proto [hunk 24]
```diff
@@ -16,6 +16,7 @@ service Api {
     rpc GetBlockByNumber(requests.GetBlockByNumberRequest) returns (responses.GetBlockByNumberResponse) {}
     rpc GetBlockHeaderByNumber(requests.GetBlockHeaderByNumberRequest) returns (responses.GetBlockHeaderByNumberResponse) {}
     rpc GetBlockInputs(requests.GetBlockInputsRequest) returns (responses.GetBlockInputsResponse) {}
+    rpc GetNoteInclusionProofs(requests.GetNoteInclusionProofsRequest) returns (responses.GetNoteInclusionProofsResponse) {}
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply proto/store.proto:19]: Yes, correct - this endpoint is only for internal usage.

File: proto/block.proto [hunk 25]
```diff
@@ -25,3 +26,8 @@ message BlockHeader {
     // the time when the block was created.
     fixed32 timestamp = 10;
 }
+
+message BlockInclusionProof {
+    BlockHeader block_header = 1;
+    merkle.MerklePath mmr_path = 2;
+}
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply proto/block.proto:33]: Not a strong preference, but it does feel odd having this message not be "self-contained". If we do want to keep it like that, we should probably add some comments explaining why. But I also don't think it is a big deal if we add a `chain_length` field here - yes, it'll be duplicated, but I think the implications of this are negligible.

File: proto/responses.proto [hunk 26]
```diff
@@ -136,6 +136,10 @@ message GetNotesByIdResponse {
     repeated note.Note notes = 1;
 }
 
+message GetNoteInclusionProofsResponse {
+    note.NoteInclusionProofs proofs = 1;
+}
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply proto/responses.proto:141]: Yeah - I think I suggested it somewhere. I think `NoteAuthenticationInfo` may be a bit better (one reason is that we use it in the context of authenticating "unauthenticated notes") - but also not a strong preference by any means.

File: crates/proto/src/domain/blocks.rs [hunk 27]
```diff
@@ -1,14 +1,14 @@
-use miden_objects::BlockHeader;
+use miden_objects::{crypto::merkle::MerklePath, BlockHeader};
 
 use crate::{
     errors::{ConversionError, MissingFieldHelper},
-    generated::block_header,
+    generated::block,
```
Comment (bobbinth) [inline crates/proto/src/domain/blocks.rs:5]: It is a bit difficult to tell what's what by looking at `block::BlockHeader` types below. I would probably alias this as:
```rust
generated::block as proto
```
And then everywhere below we could say just `proto::BlockHeader`.

File: crates/store/src/state.rs [hunk 28]
```diff
@@ -87,6 +96,43 @@ struct InnerState {
     account_tree: SimpleSmt<ACCOUNT_TREE_DEPTH>,
 }
 
+#[derive(Clone, Default, Debug)]
+pub struct NoteAuthenticationInfo {
+    pub block_proofs: Vec<BlockInclusionProof>,
+    pub note_proofs: BTreeMap<NoteId, NoteInclusionProof>,
+    pub chain_length: BlockNumber,
+}
```
Comment (bobbinth) [inline crates/store/src/state.rs:104]: Not for this PR, but this should actually be in the domain objects of `miden-proto` crate (this is because this struct is used by both store and block producer). If we do that, we won't need to add `miden-store` as a dependency to `miden-block-producer` (I think).

File: crates/block-producer/src/block_builder/mod.rs [hunk 29]
```diff
@@ -114,7 +114,9 @@ where
             .await?;
 
         let missing_notes: Vec<_> = dangling_notes
-            .difference(&block_inputs.found_unauthenticated_notes)
+            .difference(
+                &block_inputs.found_unauthenticated_notes.note_proofs.keys().copied().collect(),
```
Comment (bobbinth) [inline crates/block-producer/src/block_builder/mod.rs:118]: I would probably add a method on `NoteAuthenticationInfo` to make this a single method invocation (e.g., maybe something like `NoteAuthenticationInfo::note_ids()`).

File: crates/block-producer/src/batch_builder/tests/mod.rs [hunk 30]
```diff
@@ -260,12 +261,19 @@ async fn test_block_builder_no_missing_notes() {
 async fn test_block_builder_fails_if_notes_are_missing() {
     let accounts: Vec<_> = (1..=4).map(MockPrivateAccount::<3>::from).collect();
     let notes: Vec<_> = (1..=6).map(mock_note).collect();
+    // We require mmr for the note authentication to succeed.
+    //
+    // We also need two blocks worth of mmr because the mock store skips genesis.
+    let mut mmr = Mmr::new();
+    mmr.add(Digest::new([1u32.into(), 2u32.into(), 3u32.into(), 4u32.into()]));
+    mmr.add(Digest::new([1u32.into(), 2u32.into(), 3u32.into(), 4u32.into()]));
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply crates/block-producer/src/batch_builder/tests/mod.rs:269]: We should probably start thinking about how to refactor our testing framework.

File: crates/block-producer/src/batch_builder/batch.rs (hunk 22 above)
Comment (bobbinth) [reply crates/block-producer/src/batch_builder/batch.rs:146]: They are not currently used, but once we have batch/block kernels, they will be used in them for proving that these notes are actually authenticated.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left a few small comments inline. Once these are addressed, we can merge.

I also left one comment for a subsequent PR. I think we should do it right after this PR (it shouldn't be too complex).

File: crates/store/src/state.rs (hunk 28 above)
Comment (Mirko-von-Leipzig) [reply crates/store/src/state.rs:104]: #455

File: proto/responses.proto (hunk 26 above)
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply proto/responses.proto:141]: I've made this change -- should the proto method name change as well to `GetNoteAuthenticationInfo`?

I'm still a bit murky on the nomenclature

File: proto/block.proto (hunk 25 above)
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply proto/block.proto:33]: I've made this change; I like local reasoning.

**Comment by Mirko-von-Leipzig** [comment]
Changes made - one final naming question [here](https://github.com/0xPolygonMiden/miden-node/pull/421#discussion_r1728703554) and then we're good to go I think.

File: proto/responses.proto (hunk 26 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply proto/responses.proto]: I could go either way - so, basically, up to you :)

[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply proto/responses.proto]: changed it so that it isn't the only outlier :)
